File: client copy/src/main/java/io/github/hulang1024/chinesechess/scene/chessplay/rule/chess/ChessK.java
package io.github.hulang1024.chinesechess.scene.chessplay.rule.chess;

import io.github.hulang1024.chinesechess.scene.chessplay.rule.ChessPosition;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.HostEnum;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.RoundGame;

/**
 * 将
 * @author Hu Lang
 */
public class ChessK extends AbstractChess {
    public ChessK(ChessPosition pos, HostEnum host) {
        super(pos, host);
    }

    @Override
    public boolean canGoTo(ChessPosition destPos, RoundGame game) {
        // 可以在九宫内走单步，不限制方向（不包括斜着走）
        return MoveRules.isStraightLineMove(destPos.row - pos.row,destPos.col - pos.col, 1)
            && MoveRules.isInKingHome(this, destPos, game);
    }
}

File: web-client/chinese-chess/src/scene/play/rule/chess/ChessK.ts
import ChessHost from "../chess_host";
import ChessPos from "../ChessPos";
import RoundGame from "../RoundGame";
import AbstractChess from "./AbstractChess";
import { isInKingHome, isStraightLineMove } from "./move_rules";

/**
 * 将
 */
export default class ChessK extends AbstractChess {
    constructor(pos: ChessPos, host: ChessHost) {
        super(pos, host);
    }

    canGoTo(destPos: ChessPos, game: RoundGame) {
        // 可以在九宫内走单步，不限制方向（不包括斜着走）
        return isStraightLineMove(
                destPos.row - this.pos.row,
                destPos.col - this.pos.col,
                1)
            && isInKingHome(this, destPos, game);
    }
}
